### ISSUE TITLE
[HOLD for payment 2022-12-08] Explore cleaning up reportIOUS_ collection key

### ISSUE THREAD

**Issue by luacmartins** [Expensify/App#10525]
# Problem
Coming from [this thread](https://docs.google.com/document/d/1B_Y4T12LDl_OfTt8LMoatS8OlEwX6H162ZBxioV_xJg/edit?disco=AAAAevBC5uc), having a separate reportIOUS_ Onyx key seems unnecessary and confusing, since we create a disconnect between data by separating it into multiple different keys.

# Why is this important
Keep the code cleaner and less confusing leads to a better and faster development experience

# Solution
Explore if killing the reportIOUS_ key makes sense

**Comment by luacmartins** [comment]
On hold

**Comment by zanyrenney** [comment]
wait @luacmartins should this still be on hold? if so - why?

**Comment by luacmartins** [comment]
@zanyrenney this was part of N7: Manual Requests which is currently on hold. That being said, I'd be down to removing the hold and have this done pre Manual Requests so that we can reduce the scope of that project.

I think this should be internal though as it might involve API changes. cc @roryabraham since you showed some interest in working on this.

**Comment by roryabraham** [comment]
I definitely feel passionate that we should make this change and agree it doesn't need to be on hold, though it would probably be irresponsible of me to self-assign this right now 😞

**Comment by JmillsExpensify** [comment]
@luacmartins Did you mean to un-assign @zanyrenney?

**Comment by JmillsExpensify** [comment]
Asking because I don't think we should have any un-assigned issues in this repo anymore.

**Comment by luacmartins** [comment]
@JmillsExpensify Yes, I unassigned because this is an internal engineering job. What process should I follow for internal issues?

I also see that a bug label was added, but there's nothing wrong with the code. This is just a refactor.

**Comment by mallenexpensify** [comment]
Reassigning @zanyrenney, per the new BZ process we want all issues assigned.  Main chore SO is https://stackoverflow.com/c/expensify/questions/14418.  

BZ design doc is https://docs.google.com/document/d/14MayFMbP8g43nNnWtqtgKKnCxbmJw3v0zvk5wZaIzBE/edit#heading=h.mnjguil21u0a

**Comment by zanyrenney** [comment]
@mallenexpensify it's not clear from the link what you're requiring here. I'm OOO today so if this is urgent daily, i'll need to reassign. Please can you clarify?

**Comment by roryabraham** [comment]
This really isn't a bug – it's a refactor. Also it needs to be internal because we're changing Onyx data

**Comment by roryabraham** [comment]
My advice would be to change this to weekly too

**Comment by luacmartins** [comment]
@JmillsExpensify @mallenexpensify as Rory and I mentioned previously, this is an internal refactor and not a bug. I think that we should unassign the CM and remove the Bug label. Thoughts?

**Comment by mallenexpensify** [comment]
Based on [the BZ design doc ](https://docs.google.com/document/d/14MayFMbP8g43nNnWtqtgKKnCxbmJw3v0zvk5wZaIzBE/edit#bookmark=id.pxiu3yo8glni)
> If it’s in E/App, the issue will be auto-labeled as Internal and Weekly. This will help make clear to contributors that these issues are not available for them to make proposals on. 

So.. I think weekly and `NewFeature` makes sense.   and Zany shouldn't be assigned if it's not a bug.

**Comment by luacmartins** [comment]
Thanks for the input!

**Comment by JmillsExpensify** [comment]
Gotcha, so this is one of those refactors that is in an in between state, because parts of the Manual Requests fall within API Unchained, and then other parts fall within N7. Is that right?

**Comment by luacmartins** [comment]
>Gotcha, so this is one of those refactors that is in an in between state, because parts of the Manual Requests fall within API Unchained, and then other parts fall within N7. Is that right?

It was planned to be done with Manual Requests. However, I think there's no need to wait for that and we can go ahead with this refactor.

**Comment by JmillsExpensify** [comment]
@luacmartins Given that this requires knowledge of Manual Requests, who do you think would be a good one to pick this up then?

**Comment by JmillsExpensify** [comment]
Ok cool. This isn't a bug, and it's not strictly part of API Unchained either, so I am going to remove the daily label on this issue. Ultimately I agree, everyone in that list is working on higher value things.

**Comment by JmillsExpensify** [comment]
@Gonals not sure what else you're working on, but this could be a good cleanup item if you have any spare bandwidth.

**Comment by Gonals** [comment]
> @Gonals not sure what else you're working on, but this could be a good cleanup item if you have any spare bandwidth.

I'm a bit full this week. I'm working on the smartscan dispute handling automation and I also need to implement a Concierge Travel fix that is blocking open booking.

How urgent is this? I'll probably have some more bandwidth next week

**Comment by mountiny** [comment]
> How urgent is this? I'll probably have some more bandwidth next week

I think next week is fine, this is not very urgent.

**Comment by Gonals** [comment]
Great, I'll grab it, then!

**Comment by JmillsExpensify** [comment]
Nice, yeah not urgent.

**Comment by Gonals** [comment]
@luacmartins, @roryabraham, @marcaaron   
Hey everyone! I'm almost done with my other stuff, so I'm jumping into this.

It seems we have a fairly wide consensus to get rid of this key, but I wanted to make sure we have considered all the pros and cons!

My main concern is whether this will affect performance negatively. 

- We currently loop through all reports to create the chats, right? This change will substantially increase the number of reports in that key.

- Will `mergeCollection` suffer by increasing the size of the `Report` key?

I don't think either of this are worrisome concerns, but I wanted to raise them before jumping into implementation!

**Comment by JmillsExpensify** [comment]
What do you think about bringing that discussion to open source in Slack? Those are pretty interesting points, and we get more eyes on this there.

**Comment by marcaaron** [comment]
>We currently loop through all reports to create the chats, right? 

Curious what do we mean exactly by this? Can you link to some code about which loop in particular concerns you?

>This change will substantially increase the number of reports in that key.

"substantially increase" feels slightly hyperbolic. Number of reports will be relative to the number of chats and IOUs that you have vs. just chats. I can't really think of why this might be bad or why we should view the `reports_*` collection as limited.

>Will mergeCollection suffer by increasing the size of the Report key?

I'm not sure why it would.

----

Though, taking all that into consideration if anyone feels strongly that this doesn't need to be done right now (or at all) then nobody is forcing us to do it. The solution in the description kind of implies that there was a question mark about whether we should or not.

>Explore if killing the reportIOUS_ key makes sense

This statement could suggest that maybe we don't need to optimize for this right now:

>Keep the code cleaner and less confusing leads to a better and faster development experience

I supported the idea at one point, but not sure how valuable it really is to do.

**Comment by roryabraham** [comment]
I am still 100% on board with making this change before we start rolling out the Manual Requests in full force. It will be a nice simplification.

**Comment by Gonals** [comment]
> > We currently loop through all reports to create the chats, right?
> 
> Curious what do we mean exactly by this? Can you link to some code about which loop in particular concerns you?
> 
> > This change will substantially increase the number of reports in that key.
> 
> "substantially increase" feels slightly hyperbolic. Number of reports will be relative to the number of chats and IOUs that you have vs. just chats. I can't really think of why this might be bad or why we should view the `reports_*` collection as limited.
> 
> > Will mergeCollection suffer by increasing the size of the Report key?
> 
> I'm not sure why it would.
> 
> Though, taking all that into consideration if anyone feels strongly that this doesn't need to be done right now (or at all) then nobody is forcing us to do it. The solution in the description kind of implies that there was a question mark about whether we should or not.
> 
> > Explore if killing the reportIOUS_ key makes sense
> 
> This statement could suggest that maybe we don't need to optimize for this right now:
> 
> > Keep the code cleaner and less confusing leads to a better and faster development experience
> 
> I supported the idea at one point, but not sure how valuable it really is to do.

Coolio! I didn't feel the increase in size of the key was worrisome, but I wanted to double-check with the team just in case!

**Comment by Gonals** [comment]
> What do you think about bringing that discussion to open source or engineering-chat in Slack? Those are pretty interesting points, and we get more eyes on this there.

I think we are good. I didn't expect them to be concerns (the increase in size of the key is linear, after all), but wanted to double check with the team in case I was missing something. I have been reassured 😁

**Comment by Gonals** [comment]
I still have some cleanup left to do, but I think it is worth taking a look at [my PR](https://github.com/Expensify/App/pull/12780). 

This change, while it gets rid of the key, implies that most times we check the `Report` key for chat Reports, we need to make sure we filter out the IOUs. Are we happy with that?

**Comment by JmillsExpensify** [comment]
Looks like the Web PR is ready to merge. Woo!

**Comment by mountiny** [comment]
Assigning @mollfpr as they have reviewed the PR so we dont forget to pay them once the time comes @JmillsExpensify

**Comment by JmillsExpensify** [comment]
Ok cool. I'll have to circle back on this later as this is a new process still and I want to make sure we're aligned internally on how this should work.

**Comment by JmillsExpensify** [comment]
@mollfpr Upwork job is here: https://www.upwork.com/jobs/~018357b9f787f13d20

**Comment by JmillsExpensify** [comment]
Offer sent.

**Comment by JmillsExpensify** [comment]
Closing the loop. @mollfpr has been compensated via Upwork. @Gonals We should close automatically on a production deploy, yeah?

**Comment by Gonals** [comment]
> Assigning @mollfpr as they have reviewed the PR so we dont forget to pay them once the time comes @JmillsExpensify

Yep, although there's still a [Web cleanup issue left](https://github.com/Expensify/Expensify/issues/243532). I'll create the PR today and I'll have it ready to merge as soon as App is deployed

**Comment by JmillsExpensify** [comment]
Oh interesting. First time that's coming across my radar, thanks. Are we going to have Contributor+ review on that PR as well? If so, I'll assign myself.

**Comment by Gonals** [comment]
> Oh interesting. First time that's coming across my radar, thanks. Are we going to have Contributor+ review on that PR as well? If so, I'll assign myself.

I don't think so. That one is on the web side :)

**Comment by JmillsExpensify** [comment]
@Gonals Cool cool. We're set then. Just need to evaluate the deploy blocker above, and in addition, the PR should close out as soon as this PR hits production.

**Comment by Gonals** [comment]
We are good with the deploy blockers. It is just a console error we were already aware of (and doesn't break any functionality). The Cleanup PR should take care of it :)

**Comment by JmillsExpensify** [comment]
Ok cool. We should be good here then. Closing this one out.
